### DISCUSSION TITLE
Bootstrap JS not loading, encore

### DISCUSSION THREAD

**Discussion by plagasul** [symfony/symfony#44248] [Q&A] [answered]
Hello there. 

I am having difficulties loading Bootstrap's js with Symfony 5, encore, stimulus etc... 

jquery, popper, corejs, etc all installed and at least jquery is working... 

app.scss and app.js seems to compile well using sass, postcss and babel, including Bootstrap's CSS. Just Bootstrap's JS is not there, so no dropdowns, etc.

assets/app.js
```javascript
// any CSS you import will output into a single css file (app.css in this case)
import './styles/app.scss';

// start the Stimulus application
import './bootstrap';
```

assets/bootstrap,js
```javascript
import { startStimulusApp } from '@symfony/stimulus-bridge';

// Registers Stimulus controllers from controllers.json and in the controllers/ directory
 export const app = startStimulusApp(require.context(
    '@symfony/stimulus-bridge/lazy-controller-loader!./controllers',
    true,
    /\.(j|t)sx?$/
));

// register any custom, 3rd party controllers here
// app.register('some_controller_name', SomeImportedController);
```

note: There is no specific controller under assets/controllers

webpack.config.js
```javascript
const Encore = require('@symfony/webpack-encore');

// Manually configure the runtime environment if not already configured yet by the "encore" command.
// It's useful when you use tools that rely on webpack.config.js file.
if (!Encore.isRuntimeEnvironmentConfigured()) {
    Encore.configureRuntimeEnvironment(process.env.NODE_ENV || 'dev');
}

Encore
  // directory where compiled assets will be stored
  .setOutputPath("public/build/")
  // public path used by the web server to access the output path
  .setPublicPath("/build")
  // only needed for CDN's or sub-directory deploy
  //.setManifestKeyPrefix('build/')

  /*
   * ENTRY CONFIG
   *
   * Each entry will result in one JavaScript file (e.g. app.js)
   * and one CSS file (e.g. app.css) if your JavaScript imports CSS.
   */
  .addEntry("app", "./assets/app.js")

  // enables the Symfony UX Stimulus bridge (used in assets/bootstrap.js)
  .enableStimulusBridge('./assets/controllers.json')

  // When enabled, Webpack "splits" your files into smaller pieces for greater optimization.
  .splitEntryChunks()

  // will require an extra script tag for runtime.js
  // but, you probably want this, unless you're building a single-page app
  .enableSingleRuntimeChunk()

  /*
   * FEATURE CONFIG
   *
   * Enable & configure other features below. For a full
   * list of features, see:
   * https://symfony.com/doc/current/frontend.html#adding-more-features
   */
  .cleanupOutputBeforeBuild()
  .enableBuildNotifications()
  .enableSourceMaps(!Encore.isProduction())
  // enables hashed filenames (e.g. app.abc123.css)
  .enableVersioning(Encore.isProduction())

  .configureBabel((config) => {
    config.plugins.push("@babel/plugin-proposal-class-properties");
  })

  // enables @babel/preset-env polyfills
  .configureBabelPresetEnv((config) => {
    config.useBuiltIns = "usage";
    config.corejs = 3;
  })

  // enables Sass/SCSS support
  .enableSassLoader()

  // enable PostCSS
  .enablePostCssLoader();

// uncomment if you use TypeScript
//.enableTypeScriptLoader()

// uncomment if you use React
//.enableReactPreset()

// uncomment to get integrity="..." attributes on your script & link tags
// requires WebpackEncoreBundle 1.4 or higher
//.enableIntegrityHashes(Encore.isProduction())

// uncomment if you're having problems with a jQuery plugin
// .autoProvidejQuery()

module.exports = Encore.getWebpackConfig();
```

package.json (only dependencies)
```json
  "devDependencies": {
    "@popperjs/core": "^2.10.2",
    "@symfony/stimulus-bridge": "^2.1.0",
    "@symfony/webpack-encore": "^1.6.1",
    "autoprefixer": "^10.4.0",
    "bootstrap": "^5.1.3",
    "core-js": "^3.19.1",
    "jquery": "^3.6.0",
    "postcss-loader": "^6.2.0",
    "sass": "^1.43.4",
    "sass-loader": "^12.3.0",
    "stimulus": "^2.0.0",
    "webpack-notifier": "^1.14.1"
  },
  "dependencies": {
    "popper.js": "^1.16.1"
  }
```

How to troubleshoot further?

Thank you

**Comment by stof** [comment, marked as answer]
Bootstrap is not a Symfony UX package (it is not even a stimulus-based package) so it won't be registered automatically by `@symfony/stimulus-bridge`.

You will need to import the Bootstrap code in your app.js file as shown in the [Bootstrap documentation about how to use it with webpack](https://getbootstrap.com/docs/5.1/getting-started/webpack/)

**Reply by plagasul** [reply]
Thank you for your answer @stof .

I've been trying to see in which moment did I try to use Bootstrap with Stimulus, and I believe it was placed there by [webpack-encore-bundle](https://packagist.org/packages/symfony/webpack-encore-bundle) as I believe was suggested in Symfony Casts (I'm learning).

I am going to try and remove the bundle then install and configure encore, take it from there.

Thank you

**Reply by plagasul** [reply]
@stof Actually the bundle is reccomended in Symfonys' site to install encore: https://symfony.com/doc/current/frontend/encore/installation.html

...and I believe the bundle is what sets up Bootstrap through stimulus. If I am correct shouldn't it then just work?

Thank you

**Reply by stof** [reply]
the assets/bootstrap.js file loaded in your app.json has nothing to do with the `bootstrap` npm package. It is the file containing the bootstrapping logic for the Stimulus-based UX initiative (containing the registration of Flex-managed Symfony UX controllers and the registration of your custom Stimulus controllers)

What happens right now is that your app.js does not load Bootstrap at all, so you indeed cannot see it working (adding it in your package.json only makes it available in `node_modules`, but it won't magically be imported)

**Reply by plagasul** [reply]
@stof Thanks.

And yes, I've been confused since the beggining with Stimulus bootstrap file and importing bootstrap :facepalm: . Thank you for pointing that out.

And yes, adding `import 'bootstrap';` to app.js does seem to import  `"./node_modules/bootstrap/dist/js/bootstrap.esm.js` 

Dropdowns still not working tho. But I assume that is unrelated to Symfony at this point, and I should look elsewhere.

Thank you

**Reply by berkut1** [reply]
if you use boostrap 5, they changed options to data-bs-* for dropdowns. https://getbootstrap.com/docs/5.0/components/dropdowns/#options
P.S I spent 2 hours before I found out :)

**Reply by plagasul** [reply]
Jesus christ @berkut1 you are absolutely right, unbeliavable, I did a master on encore and symfony's architecture over the past days... and was still going at it! Thank you, really very much.
